paddle/fluid/framework/details/multi_devices_graph_builder.cc

```diff
@@ -406,10 +406,11 @@ int MultiDevSSAGraphBuilder::GetOpDeviceID(const OpDesc &op) const {
   if (op_role != static_cast<int>(framework::OpRole::kOptimize)) {
     return -1;
   }
-  std::string param = boost::get<std::string>(
+  auto &params = boost::get<std::vector<std::string>>(
       op.GetAttr(framework::OpProtoAndCheckerMaker::OpRoleVarAttrName()));
 
-  std::string grad = framework::GradVarName(param);
+  PADDLE_ENFORCE_EQ(params.size(), 1U);
+  std::string grad = framework::GradVarName(params[0]);
   int dev_id = GetVarDeviceID(grad);
```

Comment:
<img width="870" alt="qq20180712-185506 2x" src="https://user-images.githubusercontent.com/30176695/42629173-1d2bf772-8605-11e8-8d0b-f20980cf73e5.png">

In this case, sgd doesn't find its device, Because sgd's op_role_var is `fc_0.b_0`, but it's input doesnot have `fc_0.b_0`.

Reply:
It seems that we should use ` with param_and_grad[1].block.program.optimized_guard(param_and_grad[1])` here.

https://github.com/PaddlePaddle/Paddle/blob/a376efb3f301b9e5c19a119390f669ae22add3dd/python/paddle/fluid/optimizer.py#L228-L235